### PR TITLE
Improve diagnostic messages for `AssignmentStatement`, `Operation_Binary`, `ListExpression`, and `StructExpression`

### PR DESCRIPTION
**Pull request by kfcripps**

### CODE REVIEW

File: testdata/p4_14_samples_outputs/parser_pragma2.p4-stderr [hunk 1]
```diff
@@ -1,6 +1,6 @@
 parser_pragma2.p4(18): [--Wwarn=parser-transition] warning: SelectCase: unreachable
         0xAB00 : Cowles;
         ^^^^^^
-parser_pragma2.p4(16): [--Wwarn=parser-transition] warning: 'ListExpression': transition does not depend on select argument
+parser_pragma2.p4(16): [--Wwarn=parser-transition] warning: '{ tmp }': transition does not depend on select argument
```
Comment (mbudiu-vmw) [inline testdata/p4_14_samples_outputs/parser_pragma2.p4-stderr:4]: This is one problem with this improvement: this error message no longer makes sense.

File: testdata/p4_16_errors_outputs/issue2317.p4-stderr [hunk 2]
```diff
@@ -1,3 +1,3 @@
-issue2317.p4(16): [--Werror=type-error] error: ==: cannot be applied to action results
+issue2317.p4(16): [--Werror=type-error] error: f == f: cannot be applied to action results
```
Comment (mbudiu-vmw) [inline testdata/p4_16_errors_outputs/issue2317.p4-stderr:1]: this one is also confusing

**Review by mbudiu-vmw — COMMENTED** [review]
This doesn't work so well when the printed expression is not one that appeared in the source program. This is not a problem related to this PR, but this PR makes the problem a bit worse.

File: testdata/p4_14_samples_outputs/parser_pragma2.p4-stderr (hunk 1 above)
Comment (kfcripps) [reply testdata/p4_14_samples_outputs/parser_pragma2.p4-stderr:4]: I noticed that too and I am not sure of a good way to fix this. Is this problem unique to `ListExpression` components, or is it a problem when printing compiler-generated `Expression`s in other places, too?

Comment (kfcripps) [reply testdata/p4_14_samples_outputs/parser_pragma2.p4-stderr:4]: I noticed that too and I am not sure of a good way to fix this. Is this problem unique to `ListExpression` components, or is it a problem when printing compiler-generated `Expression`s in other places, too?

File: testdata/p4_16_errors_outputs/issue2317.p4-stderr (hunk 2 above)
Comment (kfcripps) [reply testdata/p4_16_errors_outputs/issue2317.p4-stderr:1]: This one can be fixed by enhancing `MethodCallExpression`'s `toString()` method. This change would affect ~110 tests so I opted to leave that for a separate PR, but I can make that change here if you prefer.
This is not unique to printing `MethodCallExpression`s used in `==` operations, though. The `MethodCallExpression` will be printed as `f` in other places too.

Comment (kfcripps) [reply testdata/p4_16_errors_outputs/issue2317.p4-stderr:1]: This one can be fixed by enhancing `MethodCallExpression`'s `toString()` method. This change would affect ~110 tests so I opted to leave that for a separate PR, but I can make that change here if you prefer.
This is not unique to printing `MethodCallExpression`s used in `==` operations, though. The `MethodCallExpression` will be printed as `f` in other places too.

File: testdata/p4_14_samples_outputs/parser_pragma2.p4-stderr (hunk 1 above)
Comment (mbudiu-vmw) [reply testdata/p4_14_samples_outputs/parser_pragma2.p4-stderr:4]: No, it happens in other places as well.

**Comment by kfcripps** [comment]
superseded by https://github.com/p4lang/p4c/pull/4353

**Pull request closed without merging**